packages/code-infra/src/utils/changelog.mjs

```diff
@@ -0,0 +1,148 @@
+import { Octokit } from '@octokit/rest';
+import { $ } from 'execa';
+
+import { persistentAuthStrategy } from './github.mjs';
+
+/**
+ * @typedef {'team' | 'first_timer' | 'contributor'} AuthorAssociation
+ */
+/**
+ * @typedef {Object} FetchedCommitDetails
+ * @property {string} sha
+ * @property {string} message
+ * @property {string[]} labels
+ * @property {number} prNumber
+ * @property {{login: string; association: AuthorAssociation} | null} author
+ */
+
+/**
+ * @param {Object} opts
+ * @param {string} opts.cwd
+ * @param {boolean} [opts.fetchAll=true] Whether to fetch all tags from all remotes before finding the latest tag.
+ * @returns {Promise<string>}
+ */
+export async function findLatestTaggedVersion(opts) {
+  const $$ = $({ cwd: opts.cwd });
+  const fetchAll = opts.fetchAll ?? true;
+  if (fetchAll) {
+    // Fetch all tags from all remotes to ensure we have the latest tags.
+    await $$`git fetch --tags --all`;
+  }
+  const { stdout } = await $$`git describe --tags --abbrev=0 --match ${'v*'}`; // only include "version-tags"
+  return stdout.trim();
+}
+
+/**
+ * @typedef {Object} FetchCommitsOptions
+ * @property {string} repo
+ * @property {string} lastRelease
+ * @property {string} release
+ * @property {string} [token]
+ * @property {string} [org="mui"]
+ */
+
+/**
+ * Fetches commits between two refs (lastRelease..release) including PR details.
+ * Automatically handles GitHub OAuth authentication.
+ *
+ * @param {FetchCommitsOptions} param0
+ * @returns {Promise<FetchedCommitDetails[]>}
+ */
+export async function fetchCommitsBetweenRefs({ org = 'mui', ...options }) {
+  const opts = { ...options, org };
+
+  return await fetchCommitsRest(opts);
+}
+
+/**
+ * Fetches commits between two refs using GitHub's REST API.
+ * It is more reliable than the GraphQL API but requires multiple network calls (1 + n).
+ * One to list all commits between the two refs and then one for each commit to get the PR details.
+ *
+ * @param {FetchCommitsOptions & { org: string }} param0
+ *
+ * @returns {Promise<FetchedCommitDetails[]>}
+ */
+async function fetchCommitsRest({ token, repo, lastRelease, release, org }) {
+  const octokit = token
+    ? new Octokit({ auth: token })
+    : new Octokit({ authStrategy: persistentAuthStrategy });
```

Comment:
Instead of a `token` would it make sense to pass through an `Octokit` instance? It would yield back all freedom to authenticate and other options to the consumer.